caproto/sync/shark.py

```diff
@@ -0,0 +1,280 @@
+    AccessRightsResponse.ID: AccessRightsResponse,
+    ClearChannelRequestOrResponse.ID: ClearChannelRequestOrResponse,
+    ClientNameRequest.ID: ClientNameRequest,
+    CreateChFailResponse.ID: CreateChFailResponse,
+    EchoRequestOrResponse.ID: EchoRequestOrResponse,
+    ErrorResponse.ID: ErrorResponse,
+    EventsOffRequest.ID: EventsOffRequest,
+    EventsOnRequest.ID: EventsOnRequest,
+    HostNameRequest.ID: HostNameRequest,
+    NotFoundResponse.ID: NotFoundResponse,
+    ReadSyncRequest.ID: ReadSyncRequest,
+    Beacon.ID: Beacon,
+    WriteRequest.ID: WriteRequest,
+    ServerDisconnResponse.ID: ServerDisconnResponse,
+    RepeaterConfirmResponse.ID: RepeaterConfirmResponse,
+    RepeaterRegisterRequest.ID: RepeaterRegisterRequest,
+}
+
+
+def sniff_version_header(header):
+    if header.parameter1 == 0:
+        return VersionRequest
+    if header.parameter1 == 1:
+        return VersionResponse
+    else:
+        raise ValueError("Unidentifiable command.")
+
+
+def sniff_search_header(header):
+    if header.payload_size == 0 or (header.payload_size == 8 and header.data_count == 0):
+        return SearchResponse
+    else:
+        return SearchRequest
+
+
+def sniff_event_add_or_cancel_header(header):
+    if header.command == 2:
+        return EventCancelRequest
+    # Special case for EventCancelResponse which is coded inconsistently.
+    if header.command == 1 and header.payload_size == 0 and header.data_count == 0:
+        return EventCancelResponse
+    if header.payload_size == 16:
+        if header.parameter1 > 60:
+            return EventAddRequest
+        else:
+            return EventAddRequestOrResponse  # probably Request but cannot be sure
+    else:
+        return EventAddResponse
+
+
+def sniff_create_chan_header(header):
+    if header.payload_size == 0:
+        return CreateChanResponse
+    else:
+        return CreateChanRequest
+
+
+def sniff_read_header(header):
+    if header.payload_size == 0:
+        return ReadRequest
+    else:
+        return ReadResponse
+
+
+def sniff_read_notify_header(header):
+    if header.payload_size == 0:
+        return ReadNotifyRequest
+    else:
+        return ReadNotifyResponse
+
+
+def sniff_write_notify_header(header):
+    if header.payload_size == 0:
+        return WriteNotifyResponse
+    else:
+        return WriteNotifyRequest
+
+
+sniffers = {
+    VersionRequest.ID: sniff_version_header,
+    SearchRequest.ID: sniff_search_header,
+    EventAddRequest.ID: sniff_event_add_or_cancel_header,
+    EventCancelRequest.ID: sniff_event_add_or_cancel_header,
+    ReadRequest.ID: sniff_read_header,
+    ReadNotifyRequest.ID: sniff_read_notify_header,
+    WriteNotifyRequest.ID: sniff_write_notify_header,
+    CreateChanRequest.ID: sniff_create_chan_header,
+}
+
+
+def infer_command_class(header):
+    id_ = header.command
+    try:
+        return one_way_commands[id_]
+    except KeyError:
+        return sniffers[id_](header)
+
+
+def shark(file):
```

Comment:
Could shark maybe made to capture non-CA packets but filter them out appropriately?

Reply:
I have been puzzling over the most reliable way to do that. Any ideas?

Reply:
Simplest would be to check each packet for something that looks like the header and try to deserialize it, ignoring ones that fail and adding them to an 'ignored' connections set (or something similar) for better performance.

Any alternative I can think of would be more complicated and fragile...

Reply:
Done. `tcpdump -w - | caproto-shark` (no source filtering) seems to work. We may discover edge cases as we use it -- things that look _enough_ like headers to trick us at first -- and the infrastructure is now in place to make validation smarter/stricter on the commands' respective `validate` meothods.